# KNN Classifier

### Dataset

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [3]:
#Load data
iris = pd.read_csv('Iris.csv')
#data cleaning
iris.drop(columns="Id",inplace=True)
iris

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm         Species
0              5.1           3.5            1.4           0.2     Iris-setosa
1              4.9           3.0            1.4           0.2     Iris-setosa
2              4.7           3.2            1.3           0.2     Iris-setosa
3              4.6           3.1            1.5           0.2     Iris-setosa
4              5.0           3.6            1.4           0.2     Iris-setosa
..             ...           ...            ...           ...             ...
145            6.7           3.0            5.2           2.3  Iris-virginica
146            6.3           2.5            5.0           1.9  Iris-virginica
147            6.5           3.0            5.2           2.0  Iris-virginica
148            6.2           3.4            5.4           2.3  Iris-virginica
149            5.9           3.0            5.1           1.8  Iris-virginica

[150 rows x 5 columns]

In [4]:
#features and labels
X=iris.iloc[:,0:4].values
y=iris.iloc[:,4].values

#Train and Test split
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

y_pred = np.zeros([len(y_test),])

# Code Starting here.......................

In [5]:
#Giving lables for flowers in dataset test dataset
for i in range(0,len(y_test)):
    if(y_test[i]=='Iris-setosa'):
        y_test[i]=1
    elif (y_test[i]=='Iris-virginica'):
        y_test[i]=2
    elif (y_test[i]=='Iris-versicolor'):
        y_test[i]=3
    else:
        print("You are a fool")


for i in range(0,len(y_train)):
    if(y_train[i]=='Iris-setosa'):
        y_train[i]=1
    elif (y_train[i]=='Iris-virginica'):
        y_train[i]=2
    elif (y_train[i]=='Iris-versicolor'):
        y_train[i]=3
    else:
        print("You are a fool")

y_train= y_train.astype('int16')
y_test= y_test.astype('int16')
y_pred= y_pred.astype('int16')
    

In [18]:
# KNN K 
KNN=5
for i in range(0,len(x_test)):
    nbrs=np.zeros([len(x_train),2])
    for j in range(0,len(x_train)):
        distance=0
        for k in range(0,len(x_train[0])):
            distance += (x_train[j][k]-x_test[i][k])**2
        nbrs[j]=[distance,y_train[j]]
    
    nbrs = nbrs[nbrs[:,0].argsort()]
    nbrs = nbrs[:KNN]
    c1=c2=c3=0;
    for e in range(0,len(nbrs)):
        if(nbrs[e][1]==1):
            c1+=1
        elif(nbrs[e][1]==2):
            c2+=1
        elif(nbrs[e][1]==3):
            c3+=1
        else:
            print("You are fooled")
    if(c1==max(c1,c2,c3)):
        y_pred[i]=1
    elif(c2==max(c1,c2,c3)):
        y_pred[i]=2
    elif(c3==max(c1,c2,c3)):
        y_pred[i]=3
    else:
        print("Fooled UP")

In [19]:
print(y_pred)
print(y_test)

[2 3 1 2 1 2 1 3 3 3 2 3 3 3 2 1 3 3 1 1 2 3 1 1 2 1 1 3 3 1]
[2 3 1 2 1 2 1 3 3 3 2 3 3 3 3 1 3 3 1 1 2 3 1 1 2 1 1 3 3 1]


In [20]:
## Accuracy Priting
import sklearn.metrics as metrics
print(np.sqrt(metrics.accuracy_score(y_test, y_pred)))


0.983192080250175


# Accuracy

## AS you can see above accuracy score is 0.98


